File: backend/app/schemas/replay.py
"""Pydantic schemas for the replay API."""

from datetime import date, datetime
from typing import Any, Dict, List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import BatchDateRange, Ticker


class ReplayRunRequest(BatchDateRange):
    model_config = ConfigDict(extra="forbid")

    scanner_type: str
    trading_strategy_id: Optional[int] = None
    universe_id: int
    max_hold_days: int = Field(default=10, ge=1, le=252)
    exit_fidelity: Literal["intraday", "daily"] = "intraday"
    benchmark_symbol: Optional[Ticker] = "SPY"


class ReplayTradeResponse(BaseModel):
    id: int
    replay_run_id: int
    scanner_event_id: Optional[int] = None
    ticker: str
    signal_date: date
    entry_date: Optional[date] = None
    entry_price: Optional[float] = None
    direction: Optional[str] = None
    stop_price: Optional[float] = None
    target_price: Optional[float] = None
    exit_date: Optional[date] = None
    exit_price: Optional[float] = None
    exit_reason: Optional[str] = None
    return_pct: Optional[float] = None
    return_r: Optional[float] = None
    mfe_pct: Optional[float] = None
    mae_pct: Optional[float] = None
    bars_held: Optional[int] = None
    regime_trend: Optional[str] = None
    regime_vol: Optional[str] = None
    fill_source: Optional[str] = None

    model_config = {"from_attributes": True}


class ReplayTradesResponse(BaseModel):
    trades: List[ReplayTradeResponse]
    total: int
    limit: int
    offset: int


class ReplayRunResponse(BaseModel):
    id: int
    uuid: UUID
    status: str
    scanner_type: str
    scanner_config_snapshot: Dict[str, Any]
    trading_strategy_id: Optional[int] = None
    strategy_snapshot: Optional[Dict[str, Any]] = None
    universe_id: int
    universe_snapshot: Dict[str, Any]
    start_date: date
    end_date: date
    max_hold_days: int
    exit_fidelity: str
    benchmark_symbol: Optional[str] = None
    data_hash: Optional[str] = None
    metrics: Optional[Dict[str, Any]] = None
    skipped_count: Optional[int] = None
    error_message: Optional[str] = None
    celery_task_id: Optional[str] = None
    total_trades: Optional[int] = None
    hit_rate: Optional[float] = None
    expectancy_r: Optional[float] = None
    profit_factor: Optional[float] = None
    max_drawdown_r: Optional[float] = None
    avg_bars_held: Optional[float] = None
    median_bars_held: Optional[float] = None
    avg_mfe_pct: Optional[float] = None
    avg_mae_pct: Optional[float] = None
    mfe_mae_ratio: Optional[float] = None
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ReplayAnalyticsResponse(BaseModel):
    status: str
    equity_curve: List[Dict[str, Any]] = []
    calendar_decay: List[Dict[str, Any]] = []
    holding_period_decay: List[Dict[str, Any]] = []
    regime_breakdown: List[Dict[str, Any]] = []


class RunCompareEntry(BaseModel):
    uuid: UUID
    scanner_type: str
    start_date: date
    end_date: date
    status: str
    headline_metrics: Dict[str, Any]
    data_hash: Optional[str]


class RunPairComparison(BaseModel):
    a: UUID
    b: UUID
    data_hash_match: bool


class ReplayCompareResponse(BaseModel):
    runs: List[RunCompareEntry]
    comparisons: List[RunPairComparison]
    all_hashes_match: bool

File: frontend/src/api/replay.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from './client';

export type ReplayRunStatus = 'queued' | 'running' | 'completed' | 'failed';
export type ReplayExitFidelity = 'intraday' | 'daily';

export interface ReplayRunCreateRequest {
  scanner_type: string;
  universe_id: number;
  start_date: string;
  end_date: string;
  trading_strategy_id?: number | null;
  max_hold_days?: number;
  exit_fidelity?: ReplayExitFidelity;
  benchmark_symbol?: string;
}

export interface ReplayRunSummary {
  id: number;
  uuid: string;
  status: ReplayRunStatus;
  celery_task_id: string | null;
  scanner_type: string;
  trading_strategy_id: number | null;
  universe_id: number;
  start_date: string;
  end_date: string;
  max_hold_days: number;
  exit_fidelity: ReplayExitFidelity;
  benchmark_symbol: string;
  data_hash: string | null;
  metrics: ReplayMetrics | null;
  skipped_count: number;
  error_message: string | null;
  total_trades: number;
  hit_rate: number | null;
  expectancy_r: number | null;
  profit_factor: number | null;
  max_drawdown_r: number | null;
  avg_bars_held: number | null;
  median_bars_held: number | null;
  avg_mfe_pct: number | null;
  avg_mae_pct: number | null;
  mfe_mae_ratio: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface ReplayTrade {
  id: number;
  ticker: string;
  signal_date: string;
  entry_date: string | null;
  entry_price: number | null;
  direction: 'long' | 'short';
  stop_price: number | null;
  target_price: number | null;
  exit_date: string | null;
  exit_price: number | null;
  exit_reason: string | null;
  return_pct: number | null;
  return_r: number | null;
  mfe_pct: number | null;
  mae_pct: number | null;
  bars_held: number | null;
  regime_trend: string | null;
  regime_vol: string | null;
  fill_source: string | null;
}

export interface ReplayTradesResponse {
  trades: ReplayTrade[];
  total: number;
  limit: number;
  offset: number;
}

export interface ReplayMetrics {
  equity_curve?: Array<{ trade_index: number; equity_r: number }>;
  calendar_decay?: Array<{ bucket: string; trades: number; expectancy_r: number | null }>;
  holding_period_decay?: Array<{ bars_held: number; trades: number; expectancy_r: number | null }>;
  regime_breakdown?: Array<{
    trend: string;
    volatility: string;
    trades: number;
    expectancy_r: number | null;
    hit_rate: number | null;
  }>;
  [key: string]: unknown;
}

export interface ReplayCompareResponse {
  runs: ReplayRunSummary[];
  comparisons: Array<{
    a: string;
    b: string;
    data_hash_match: boolean;
  }>;
}

export interface ReplayRunListParams {
  scanner_type?: string;
  status?: ReplayRunStatus;
  universe_id?: number;
  limit?: number;
  offset?: number;
}

export const fetchReplayRuns = async (params?: ReplayRunListParams): Promise<ReplayRunSummary[]> => {
  const response = await apiClient.get('/replay/runs', { params });
  return response.data;
};

export const fetchReplayRun = async (uuid: string): Promise<ReplayRunSummary> => {
  const response = await apiClient.get(`/replay/runs/${uuid}`);
  return response.data;
};

export const createReplayRun = async (payload: ReplayRunCreateRequest): Promise<ReplayRunSummary> => {
  const response = await apiClient.post('/replay/runs', payload);
  return response.data;
};

export const fetchReplayTrades = async (
  runUuid: string,
  params?: { limit?: number; offset?: number; sort?: string; direction?: 'asc' | 'desc' },
): Promise<ReplayTradesResponse> => {
  const response = await apiClient.get(`/replay/runs/${runUuid}/trades`, { params });
  return response.data;
};

export const fetchReplayAnalytics = async (runUuid: string): Promise<ReplayMetrics> => {
  const response = await apiClient.get(`/replay/runs/${runUuid}/analytics`);
  return response.data;
};

export const compareReplayRuns = async (ids: string[]): Promise<ReplayCompareResponse> => {
  const response = await apiClient.get('/replay/runs/compare', { params: { ids: ids.join(',') } });
  return response.data;
};

export const useReplayRuns = (params?: ReplayRunListParams) =>
  useQuery({
    queryKey: ['replay', 'runs', params],
    queryFn: () => fetchReplayRuns(params),
    refetchInterval: 15_000,
  });

export const useReplayRun = (uuid?: string | null) =>
  useQuery({
    queryKey: ['replay', 'run', uuid],
    queryFn: () => fetchReplayRun(uuid as string),
    enabled: Boolean(uuid),
    refetchInterval: (query) => {
      const status = query.state.data?.status;
      return status === 'queued' || status === 'running' ? 5_000 : false;
    },
  });

export const useReplayTrades = (runUuid?: string | null, sort = 'signal_date', direction: 'asc' | 'desc' = 'desc') =>
  useQuery({
    queryKey: ['replay', 'trades', runUuid, sort, direction],
    queryFn: () => fetchReplayTrades(runUuid as string, { limit: 250, sort, direction }),
    enabled: Boolean(runUuid),
  });

export const useReplayAnalytics = (runUuid?: string | null) =>
  useQuery({
    queryKey: ['replay', 'analytics', runUuid],
    queryFn: () => fetchReplayAnalytics(runUuid as string),
    enabled: Boolean(runUuid),
  });

export const useCreateReplayRun = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: createReplayRun,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['replay', 'runs'] });
    },
  });
};

export const useCompareReplayRuns = () =>
  useMutation({
    mutationFn: compareReplayRuns,
  });
